### ISSUE TITLE
DNS over TLS - Add additional mode - TLS without strict Certificate enforcement, no downgrade

### ISSUE THREAD

**Issue by shoulders** [systemd/systemd#33807]
### Component

systemd-resolved

### Is your feature request related to a problem? Please describe

There are 3 modes currently for DNS over TLS (DoT)

- no = non encrypted only
- yes = Strict mode 
- opportunistic = Will use TLS if it finds it, is not bothered about certificates but will downgrade to non-encrypted if it cannot successfully do aDNS lookup with TLS

### My setup
- On my local network behind a NAT I have a webserver that handles all of my DNS and website traffic.
- I have a pfSense router that handles all of my DNS and network traffic.
- my pfsense is set to hijack all DNS traffic to force it to go through the DNS server I specify.
- My webserver is a Ubunut 22.04 server with systemd installed.
- my pfSense router is my DNS server.

### The problem
- i dont want to go through the hassle of setting up scripts to move certificates from my webserver to my pfsense box so I can use `DNSOverTLS=yes` but I do not want DNS requests from my webserver to my pfSense router to go unencrypted.
- I can use `DNSOverTLS=opportunistic` but if there are issues the connectino can be downgraded to non-encrypted.

### Describe the solution you'd like

I would alike an additional mode for `DNSOverTLS=` which will only allo TLS connections, cannot be downgraded to non-encrypted but is not strict with the certificate validation.

This will allow my webserver's DNS requests to always go in TLS to my pfSense router but with a valid SSL certificate or just a self-signed certificate.

### Describe alternatives you've considered

opportunistic mode but it will downgrade to non encrypted

### The systemd version you checked that didn't have the feature you are asking for

249

**Comment by pemensik** [comment]
I think this is wrong idea. Instead, you should be able to add accepted certificate, which does not need to match server name. Or additional CA used for DNS only. Without any checking of ceritificate, whole TLS channel is pointless and brings no real security.

Have been tried adding self-signed certificate into your accepted CA authorities? That should allow your custom certificate, but still protect you against other attackers.

But, hijacking other TLS servers should not be done in any case. If you need to forbid them, why don't you block their addresses instead? So user either accepts your server or has no working resolution on your network?

**Comment by shoulders** [comment]
@pemensik I think you have missed the whole point of my request so I will try again 😄 

I want `opportunistic` mode without fallback to unencrypted DNS on port 53.

- this is more secure than `opportunistic` mode but less secure than `Strict` mode.
- this will prevent me having to do over complicated CA and certificate solutions on my local network for one server's client lookups.

**Comment by pemensik** [comment]
No, I understand your request @shoulders , but think that way is _a bad_ idea. I think that ignoring CA chain in ``opportunistic`` mode is wrong _also_. It may ignore hostname match (we do not know correct name for the server), but should not ignore checking trusted CA signature or whitelist status. It should expose accepted name in ``resolvectl status`` or similar command, if we are accepting any server name.

I think (safer) solution to your issue is simpler whitelist of extra certificates accepted, perhaps similar way as positive dnssec trust anchor addition. Just make it possible to accept your server only, not any server wherever you go. If you intercept https traffic also, you may want to add that self-signed cert into CA set for your machines. You do a lot of complicated stuff already. Creating your custom CA should not be so hard on top of it.

When anyone can provide TLS server endpoint with any certificate, then there is no point in wasting CPU cycles on encryption. It provides just false sense of security that way. XORing your data by random value would provide similar level of protection. It may be safer than ``opportunistic``, but just a little bit. There are better ways to make it reasonably safe.

**Comment by shoulders** [comment]
@pemensik thanks for the generous responses

I think you are over engineering a solution to a problem that does not exist. I am not a corporation with many man hours to configure certificates here, certificates there, I have a about 5 computers behind a NAT.

`opportunistic mode` exist for circumstances like mine, period, if it was that bad it should never of been added yet it was. My solution adds to this modes security.

I am not going to learn all of these methods you have mentioned as for me it is pointless and a waste of my time, they will add no extra security at all. All my DNS traffic is hijacked by my pfSense router so all of my DNS traffic goes to my router (DoH, DoT, DNS, DoQ). The only DNS servers I have configured on my Unbuntu server, is my pfSense router. I just want to encrypt the connection between the two devices which is currently happening because I use `opportunistic mode` but with a chance of fallback to DNS 53 (which I don't want)

Additionally, when TLS1.3 is released there will be a performance increase for my lookups by using my proposed mode otherwise I would be restricted to slow connections.

> I think that ignoring CA chain in opportunistic mode is wrong also. 

- This is out of scope of this request.
- if `opportunistic mode` is removed then I would be forced to use non-encrypted, how is that good.
- I don't care about server validation (in this case), I just want a TLS pipe to send my packets over which is all on my local network.

> When anyone can provide TLS server endpoint with any certificate, then there is no point in wasting CPU cycles on encryption.

- wrong
- A local network behind a NAT where you cannot issue TLS certificates to all devices.
- prevents MITM attacks on DNS in a compromised local network
- Prevents unwanted apps sniffing my traffic
- The option has to be selected by the admin, I am not asking for it to be on by default and I understand the ramifications if I select this.
